src/gui/fspathedit.cpp

```diff
@@ -0,0 +1,365 @@
+    return d->m_mode;
+}
+
+void FileSystemPathEdit::setMode(FileSystemPathEdit::Mode theMode)
+{
+    Q_D(FileSystemPathEdit);
+    d->m_mode = theMode;
+    d->modeChanged();
+}
+
+QString FileSystemPathEdit::dialogCaption() const
+{
+    Q_D(const FileSystemPathEdit);
+    return d->m_dialogCaption;
+}
+
+void FileSystemPathEdit::setDialogCaption(const QString &caption)
+{
+    Q_D(FileSystemPathEdit);
+    d->m_dialogCaption = caption;
+}
+
+QString FileSystemPathEdit::currentDirectory() const
+{
+    return QFileInfo(selectedPath()).absoluteDir().absolutePath();
+}
+
+QWidget *FileSystemPathEdit::editWidgetImpl() const
+{
+    Q_D(const FileSystemPathEdit);
+    return d->m_editor->widget();
+}
+
+// ------------------------- FileSystemPathLineEdit ----------------------
+FileSystemPathLineEdit::FileSystemPathLineEdit(QWidget *parent)
+    : FileSystemPathEdit(new WidgetType(), parent)
+{
+    connect(editWidget<WidgetType>(), SIGNAL(editingFinished()), this, SLOT(onPathEdited()));
+    connect(editWidget<WidgetType>(), SIGNAL(textChanged(QString)), this, SLOT(onPathEdited()));
+}
+
+QString FileSystemPathLineEdit::editWidgetText() const
+{
+    return editWidget<WidgetType>()->text();
+}
+
+void FileSystemPathLineEdit::clear()
+{
+    editWidget<WidgetType>()->clear();
+}
+
+void FileSystemPathLineEdit::setEditWidgetText(const QString &text)
+{
+    editWidget<WidgetType>()->setText(text);
+}
+
+// ----------------------- FileSystemPathComboEdit -----------------------
+FileSystemPathComboEdit::FileSystemPathComboEdit(QWidget *parent)
+    : FileSystemPathEdit(new WidgetType(), parent)
+{
+    editWidget<WidgetType>()->setEditable(true);
+    connect(editWidget<WidgetType>(), SIGNAL(currentTextChanged(const QString&)), this, SLOT(onPathEdited()));
+    connect(editWidget<WidgetType>()->lineEdit(), SIGNAL(editingFinished()), this, SLOT(onPathEdited()));
+}
+
+void FileSystemPathComboEdit::clear()
+{
+    editWidget<WidgetType>()->clear();
+}
+
+int FileSystemPathComboEdit::count() const
+{
+    return editWidget<WidgetType>()->count();
+}
+
+QString FileSystemPathComboEdit::item(int index) const
+{
+    return Utils::Fs::fromNativePath(editWidget<WidgetType>()->itemText(index));
+}
+
+void FileSystemPathComboEdit::addItem(const QString &text)
+{
+    editWidget<WidgetType>()->addItem(Utils::Fs::toNativePath(text));
+}
+
+int FileSystemPathComboEdit::currentIndex() const
+{
+    return editWidget<WidgetType>()->currentIndex();
+}
+
+void FileSystemPathComboEdit::setCurrentIndex(int index)
+{
+    editWidget<WidgetType>()->setCurrentIndex(index);
+}
+
+QString FileSystemPathComboEdit::editWidgetText() const
+{
+    return editWidget<WidgetType>()->currentText();
+}
+
+void FileSystemPathComboEdit::setEditWidgetText(const QString &text)
+{
+#ifdef QBT_USES_QT5
+    editWidget<WidgetType>()->setCurrentText(text);
+#else
+    editWidget<WidgetType>()->lineEdit()->setText(text);
+#endif
+}
+
+#include "moc_fspathedit.cpp"
```

Comment:
Due to  `Q_PRIVATE_SLOT(d_func(), void browseActionTriggered())` in the `FileSystemPathEdit` class.